sys/include/tsrb.h

```diff
@@ -0,0 +1,154 @@
+/*
+ * Copyright (C) 2015 Kaspar Schleiser <kaspar@schleiser.de>
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @defgroup    sys_tsrb Thread safe ringbuffer
+ * @ingroup     sys
+ * @{
+ */
+
+/**
+ * @file
+ * @brief       Thread-safe ringbuffer implementation
+ *
+ * This ringbuffer implementation can be used without locking if
+ * there's only one producer and one consumer.
+ *
+ * @note Buffer size must be a power of two!
+ *
+ * @author      Kaspar Schleiser <kaspar@schleiser.de>
+ */
+
+#ifndef TSRB_H
+#define TSRB_H
+
+#include <assert.h>
+#include <stddef.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @brief     thread-safe ringbuffer struct
+ */
+typedef struct tsrb {
+    char *buf;                  /**< Buffer to operate on. */
+    unsigned int size;          /**< Size of buf. */
+    volatile unsigned reads;    /**< total number of reads */
+    volatile unsigned writes;   /**< total number of writes */
+} tsrb_t;
+
+/**
+ * @brief Static initializer
+ */
+#define TSRB_INIT(BUF) { (BUF), sizeof (BUF), 0, 0 }
+
+/**
+ * @brief        Initialize a tsrb.
+ * @param[out]   rb        Datum to initialize.
+ * @param[in]    buffer    Buffer to use by tsrb.
+ * @param[in]    bufsize   `sizeof (buffer)`
+ */
+static inline void tsrb_init(tsrb_t *rb, char *buffer, unsigned bufsize)
+{
+    /* make sure bufsize is a power of two.
+     * http://www.exploringbinary.com/ten-ways-to-check-if-an-integer-is-a-power-of-two-in-c/
+     */
+    assert((bufsize != 0) && ((bufsize & (~bufsize + 1)) == bufsize));
+
+    rb->buf = buffer;
+    rb->size = bufsize;
+    rb->reads = 0;
+    rb->writes = 0;
+}
+
+/**
+ * @brief       Test if the tsrb is empty.
+ * @param[in]   rb  Ringbuffer to operate on
+ * @return      0   if not empty
+ * @return      1   otherwise
+ */
+static inline int tsrb_empty(const tsrb_t *rb)
+{
+    return (rb->reads == rb->writes);
+}
+
+
+/**
+ * @brief       Get number of bytes available for reading
+ * @param[in]   rb  Ringbuffer to operate on
+ * @return      nr of available bytes
+ */
+static inline unsigned int tsrb_avail(const tsrb_t *rb)
+{
+    return (rb->writes - rb->reads);
+}
+
+/**
+ * @brief       Test if the tsrb is full
+ * @param[in]   rb  Ringbuffer to operate on
+ * @return      0   if not full
+ * @return      1   otherwise
+ */
+static inline int tsrb_full(const tsrb_t *rb)
+{
+    return (rb->writes - rb->reads) == rb->size;
```

Comment:
this is just a matter of taste and should not be a blocker for me, but: what's your opinion on replacing `(rb->writes - rb->reads)` with `tsrb_avail(rb)` and in line 111 also `rb->writes + rb->reads` replacing with `tsrb_avail(rb)`?